### PR TITLE
check both stores for an action while migration is in progress

### LINKED ISSUE #486: Uncaught InvalidArgumentException in system status
We have a couple of similar reports where code is complaining about action id not being valid, e.g.

```

2020-03-11T09:58:16+00:00 CRITICAL Uncaught InvalidArgumentException: Ongeïdentificeerde actie 3 in /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php:495
Stack trace:
#0 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php(482): ActionScheduler_wpPostStore->get_date_gmt('3')
#1 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_HybridStore.php(274): ActionScheduler_wpPostStore->get_date('3')
#2 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_wcSystemStatus.php(78): ActionScheduler_HybridStore->get_date('3')
#3 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_wcSystemStatus.php(52): ActionSchedu in /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php op de lijn 495

2020-03-11T10:00:08+00:00 CRITICAL Uncaught InvalidArgumentException: Ongeïdentificeerde actie 3 in /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php:495
Stack trace:
#0 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php(482): ActionScheduler_wpPostStore->get_date_gmt('3')
#1 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_HybridStore.php(274): ActionScheduler_wpPostStore->get_date('3')
#2 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_wcSystemStatus.php(78): ActionScheduler_HybridStore->get_date('3')
#3 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_wcSystemStatus.php(52): ActionSchedu in /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php op de lijn 495

2020-03-11T10:00:23+00:00 CRITICAL Uncaught InvalidArgumentException: Ongeldig actie ID. Geen status gevonden. in /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php:740
Stack trace:
#0 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_HybridStore.php(326): ActionScheduler_wpPostStore->get_status('3')
#1 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_ListTable.php(533): ActionScheduler_HybridStore->get_status('3')
#2 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/abstracts/ActionScheduler_Abstract_ListTable.php(659): ActionScheduler_ListTable->prepare_items()
#3 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_AdminView.php(94): Act in /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php op de lijn 740

2020-03-11T10:02:33+00:00 CRITICAL Uncaught InvalidArgumentException: Ongeïdentificeerde actie 3 in /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php:495
Stack trace:
#0 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php(482): ActionScheduler_wpPostStore->get_date_gmt('3')
#1 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_HybridStore.php(274): ActionScheduler_wpPostStore->get_date('3')
#2 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_wcSystemStatus.php(78): ActionScheduler_HybridStore->get_date('3')
#3 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_wcSystemStatus.php(52): ActionSchedu in /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php op de lijn 495

2020-03-11T10:03:57+00:00 CRITICAL Uncaught InvalidArgumentException: Ongeïdentificeerde actie 177 in /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php:457
Stack trace:
#0 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_HybridStore.php(248): ActionScheduler_wpPostStore->cancel_action('177')
#1 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/abstracts/ActionScheduler_Store.php(290): ActionScheduler_HybridStore->cancel_action('177')
#2 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/abstracts/ActionScheduler_Store.php(275): ActionScheduler_Store->bulk_cancel_actions(Array)
#3 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/woocommerce-admin/src/Schedulers/SchedulerTraits.php(356): Actio in /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php op de lijn 457

2020-03-11T10:04:10+00:00 CRITICAL Uncaught InvalidArgumentException: Ongeïdentificeerde actie 3 in /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php:495
Stack trace:
#0 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php(482): ActionScheduler_wpPostStore->get_date_gmt('3')
#1 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_HybridStore.php(274): ActionScheduler_wpPostStore->get_date('3')
#2 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_wcSystemStatus.php(78): ActionScheduler_HybridStore->get_date('3')
#3 /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_wcSystemStatus.php(52): ActionSchedu in /mnt/web419/b0/82/59534982/htdocs/WordPress_03/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_wpPostStore.php op de lijn 495

```

[report 1](https://wordpress.org/support/topic/fatal-error-in-scheduler/)
[report 2](https://wordpress.org/support/topic/action-scheduler-critical-uncaught-invalidargumentexception/)

What I find a bit perplexing is that (by having a quick look), this code wants to get information about actions with relatively low post ids, which seems suspicious (post id 3 or 2 in these cases).

### PR DESCRIPTION
**Pull request by rrennick**
Closes #486 

This PR updates the Hybrid data store to use the primary & secondary data stores' `fetch_action` to determine whether an action ID exists in that store before using the store to retrieve action data. Both data stores implement exception handling in `fetch_action` and return a NullAction action when the action is not found/valid.

###

- Determine the minimum action ID in the tables data store `SELECT MIN(action_id) FROM wp_actionscheduler_actions;`
- Ensure your migration is complete
- Create a number of pending actions
```
add_action( 'init', function() {
	for ( $i = 1; $i < 200; $i++ )
	as_enqueue_async_action( 'as-async-test' . $i );
} );
```
- Add a log to check that actions are being processed before the migration action is processed
```
add_action( 'action_scheduler_after_execute', function( $action_id, $action ) {
	error_log( $action->get_hook() );
}, 10, 2 );
```
- Update your demarkation ID to be 10000 greater than the maximum `action_id` in the actions table
`wp option update action_scheduler_hybrid_store_demarkation 123456` 
- Reset the migration complete flag `wp option delete action_scheduler_migration_status`
- Load the Tools -> Scheduled Actions screen
- Check the `Pending`/`Completed` list for the test actions to be processed
- Migration should complete
- No errors or warnings should be logged

### CODE REVIEW

**Comment by rrennick** [comment]
@thenbrent Could you give this one a sanity check where I added `null` return values?

**Comment by thenbrent** [comment]
This is a bit of a dangerous change, but it looks good @rrennick. I like the introduction of a central `get_store_from_action_id()` method.

The reason I says it's a dangerous changes is because prior to this patch, if an action had an `$action_id < $this->demarkation_id`, it had to exist in the old, post data store for it to be run. Now it can exist in either and still be run.

Processing actions under new conditions creates the risk of processing an action when it shouldn't be run. I can't see any conditions where this would cause a problem. Even when considering corner-cases like upgrade-downgrade-upgrade. However, I want to mention the risk as it should be considered before merging.

**Comment by becdetat** [comment]
I'm unqualified to talk about the risk but I followed the instructions and I'm pretty sure it tested ok.

**Comment by rrennick** [comment]
> Now it can exist in either and still be run.

@thenbrent Thanks, good point.

@becdetat Thanks for the feedback :)

**Comment by rrennick** [comment]
@jeffstieler This is ready for a review.

File: classes/data-stores/ActionScheduler_HybridStore.php [hunk 1]
```diff
@@ -324,13 +321,45 @@ public function mark_complete( $action_id ) {
...
+	 * @return ActionScheduler_Store
+	 */
+	protected function get_store_from_action_id( $action_id, $primary_first = false ) {
 		if ( $action_id < $this->demarkation_id ) {
-			return $this->secondary_store->get_status( $action_id );
+			$stores = [
+				$this->primary_store,
+				$this->secondary_store,
+			];
+		} elseif ( $action_id < $this->demarkation_id ) {
```
Comment (jeffstieler) [inline classes/data-stores/ActionScheduler_HybridStore.php:344]: This looks to be the same condition as the initial `if()` - typo?

File: classes/data-stores/ActionScheduler_HybridStore.php [hunk 2]
```diff
@@ -324,13 +321,45 @@ public function mark_complete( $action_id ) {
...
+	 * @param bool $primary_first Optional flag indicating search the primary store first.
+	 * @return ActionScheduler_Store
+	 */
+	protected function get_store_from_action_id( $action_id, $primary_first = false ) {
 		if ( $action_id < $this->demarkation_id ) {
-			return $this->secondary_store->get_status( $action_id );
+			$stores = [
+				$this->primary_store,
+				$this->secondary_store,
+			];
```
Comment (jeffstieler) [inline classes/data-stores/ActionScheduler_HybridStore.php:343]: Should this order be reversed to match the original logic - or is the order change intentional?

File: classes/data-stores/ActionScheduler_HybridStore.php (hunk 1 above)
Comment (rrennick) [reply classes/data-stores/ActionScheduler_HybridStore.php:344]: @jeffstieler Thanks for catching that. I updated the commit.

File: classes/data-stores/ActionScheduler_HybridStore.php (hunk 2 above)
Comment (rrennick) [reply classes/data-stores/ActionScheduler_HybridStore.php:343]: This is intentional.

**Comment by rrennick** [comment]
@jeffstieler This is ready for another review.

**Pull request merged**